src/stream/src/executor/backfill/snapshot_backfill/executor.rs

```diff
@@ -414,13 +453,14 @@ struct UpstreamBuffer<'a, S> {
     upstream: &'a mut MergeExecutorInput,
     max_pending_epoch_lag: u64,
     consumed_epoch: u64,
-    pending_non_checkpoint_barriers: Vec<DispatcherBarrier>,
+    // newer barrier at the front
+    pending_non_checkpoint_barriers: VecDeque<DispatcherBarrier>,
     /// Barriers received from upstream but not yet received the barrier from local barrier worker.
     ///
     /// In the outer `VecDeque`, newer barriers at the front.
-    /// In the inner `Vec`, newer barrier at the back, with the last barrier as checkpoint barrier,
+    /// In the inner `VecDeque`, newer barrier at the front, with the last barrier as checkpoint barrier,
     /// and others as non-checkpoint barrier
-    upstream_pending_barriers: VecDeque<Vec<DispatcherBarrier>>,
+    upstream_pending_barriers: VecDeque<VecDeque<DispatcherBarrier>>,
```

Comment:
Why do we have 2 layers to VecDeque, could you elaborate on this data structure?

Reply:
You may refer to the rust doc of the field.